freedomofpress/securedrop-workstation
# Permit whitelisting VMs for copy/paste & copying logs via tags

Adds RPC policies that let tagged VMs send clipboard pastes to `sd-app`, receive clipboard pastes from `sd-app`, and receive files from `sd-log`.

## Status

Ready for review; docs in progress at https://github.com/freedomofpress/securedrop-workstation-docs/pull/35 and will need to be kept in sync with changes here.

Adds RPC policies for the following tags:
- `sd-receive-app-clipboard` -- VMs with this tag will be able to receive clipboard contents from `sd-app`
- `sd-send-app-clipboard` -- VMs with this tag will be able to send clipboard contents to `sd-app`
- `sd-receive-logs` -- VMs with this tags will be able to receive any file sent via `qvm-copy` from `sd-log`

Adds a script in `scripts/remove-tags` that removes all tags, and runs it for `make clean` / `securedrop-admin --uninstall` operations.

## Description of Changes

Fixes #526
Towards #529

## Testing

### Scenario: Basic use

#### Prerequisites

1. Check out the version of `sd-dom0-qvm-rpc.sls` from this branch and apply it to your `/srv/salt` directory in an up-to-date production or staging install, or provision a dev environment using `make dev` from this branch.
2. For production or staging environments, run `securedrop-admin --apply` to apply the change.

#### Test
1. Add the ability to receive clipboard contents from `sd-app` to a *target VM*, e.g. `qvm-tags work add sd-receive-app-clipboard`  (`qvm-tags <target VM name> add sd-receive-app-clipboard`)
2. Add the ability to send clipboard contents from a single *source VM* to `sd-app`, e.g. `qvm-tags vault add sd-send-app-clipboard` (`qvm-tags <source VM name> add sd-send-app-clipboard`).
3. Add the ability to receive logs to a *target VM*, e.g., `qvm-tags work add sd-receive-logs`
4. Launch the client
5. - [ ] Observe that you are able to copy clipboard contents from the SecureDrop Client reply box to the *target VM* specified in step 1
6. - [ ] Observe that you are able to copy clipboard contents from the *source VM* specified in step 2 to an input field in the SecureDrop Client
7. - [ ] Observe that you are able to copy files from `sd-log` to the *target VM* specified in step 3 using `qvm-copy-to-vm` and `qvm-copy`.
8. - [ ] Observe that you are not able to perform copy/paste from any SecureDrop VM to any other VM, except from `sd-app` to *target VM* you configured in step 1
9. - [ ] Observe that you are not able to perform copy/paste to any SecureDrop VM from any VM except from the *source VM* you configured in step 4 to `sd-app`.
10. - [ ] Observe that you are not able to perform qvm-copy *to* any SecureDrop VMs except from `sd-proxy` to `sd-app`.
11. - [ ] Observe that you are not able to perform qvm-copy *from* any SecureDrop VMs except for `sd-log` as permitted by the `sd-log` rules you configured.
12. Remove the tags by running `scripts/remove-tags` from this branch.
13. - [ ] Observe via `qvm-tags <name of VM> ls` that tags were correctly removed.
14. - [ ] Observe that the copy & paste rules you configured have been disabled.
15. - [ ] Observe that the `sd-log` rules you configured have been disabled.
16. - [ ] Spot check that the rules otherwise still behave as before.

### Scenario: Uninstall (prod)

#### Prerequisites
1. You must have a working staging or prod install you're willing to lose :cry: 
2. Re-tag select VMs as per the test plan above.
3. Make sure that the changes in this PR are applied to the `/bin` version of `securedrop-admin`, and to the files in `/srv/salt` touched by this PR (create `remove-tags`, update `sd-clean-all.sls`)
#### Test
1. Uninstall via `securedrop-admin --uninstall`.
2. - [ ] Observe that the tags you added are correctly removed as part of the uninstall.

### Scenario: Uninstall (dev)
#### Prerequisites

1. You must have a working dev install you are willing to lose :cry: 
2. Re-tag select VMs as per the test plan above.
3. Ensure that your `securedrop-workstation` checkout directory in `dom0` is using this branch.

#### Test
1. Uninstall via `make clean`
2. - [ ] Observe that the tags are correctly removed as part of the uninstall.

## Checklist

- [x] RPM updated to include new files (release not bumped yet)
- [ ] `make test` runs fine in `dom0` (not tested yet; test for RPC policies has been updated, however).
